### DISCUSSION TITLE
Writing some client code for a pygls server - `textDocument/completion` isnt a recognized method?

### DISCUSSION THREAD

**Discussion by blankRiot96** [openlawlibrary/pygls#404] [Q&A]
Say I have a python file, `D:\d\p\editor\main.py`
```py
from src.core import Core


def main():
    core = Core()
    core.run()


x = m

if __name__ == "__main__":
    main()
```
And at the position `9, 6` (the position after the character `m`) I wanted to fetch some autocompletion suggestions

Well, I read the docs regarding the features that have been implemented, this section:
![image](https://github.com/openlawlibrary/pygls/assets/77634274/0f5ab27b-e3fe-4ead-9993-6611eea2586c)

But I can't really find what the request message should look like in order to *use* `textDocument/completion`?

Here's what I had for the server
```py
from pygls.server import LanguageServer

server = LanguageServer("example-server", "v0.1")

server.start_ws("127.0.0.1", 2087)

```

And for the client:
```py
import json

import websocket

LSP_WS_URL = "ws://127.0.0.1:2087"

completion_request = {
    "jsonrpc": "2.0",
    "id": 1,
    "method": "textDocument/completion",
    "params": {
        "textDocument": {"uri": "file:///D:/d/p/editor/main.py"},
        "position": {
            "line": 9,
            "character": 6,
        },
    },
}

ws = websocket.create_connection(LSP_WS_URL)
ws.send(json.dumps(completion_request))
response = ws.recv()
ws.close()

response_data = json.loads(response)
suggestions = response_data.get("result", [])

for item in suggestions:
    print(item["label"])
```

However, this is where i get the error
```
    raise JsonRpcMethodNotFound.of(feature_name)
pygls.exceptions.JsonRpcMethodNotFound: Method Not Found: textDocument/completion

```

Also, is there a better/easier way to write client code? I dont mind websockets but maybe yall have some suggestions

**Comment by blankRiot96** [comment]
Have I severely misunderstood what pygls does? I think so. So it doesn't provide any default completion engines for suggesting autocompletion?

**Reply by alcarney** [reply]
Ah apologies, I can see how the documentation can be a bit misleading. :/

`pygls` is a library/framework for writing your own language server implementation for a given language, so it doesn't offer any completion suggestions out of the box. Your screenshot above is referring to an [example server implementation](https://github.com/openlawlibrary/pygls/blob/8e1e8fa3b1ab16fcb804f7d5330ece2e5583206b/examples/servers/json_server.py) we have for JSON files.

If you are looking for a language server that supports Python you might be interested in [`jedi-language-server`](https://github.com/pappasam/jedi-language-server) (which uses `pygls`) or [`pyright`](https://microsoft.github.io/pyright/#/) (written in TypeScript).

As for writing client code, `pygls` currently offers a [`BaseLanguageClient`](https://github.com/openlawlibrary/pygls/blob/8e1e8fa3b1ab16fcb804f7d5330ece2e5583206b/pygls/lsp/client.py) class which you can base your implementation on. It's fairly new so it's not well documented (yet!) but supports communicating with a server over stdin/stdout if you are looking for an alternative to websockets.

- [Here](https://github.com/openlawlibrary/pygls/blob/8e1e8fa3b1ab16fcb804f7d5330ece2e5583206b/tests/lsp/test_completion.py) is an example of making a `textDocument/completion` request.
- [Here](https://github.com/openlawlibrary/pygls/blob/8e1e8fa3b1ab16fcb804f7d5330ece2e5583206b/tests/client.py#L87) is where that test client is defined and [how we start the server](https://github.com/openlawlibrary/pygls/blob/8e1e8fa3b1ab16fcb804f7d5330ece2e5583206b/tests/client.py#L162)

Hopefully that clears up some of the confusion and I'm happy to help if you have any more questions :)

**Comment by blankRiot96** [comment]
Yeah I found jedi later on, thank you for your reply
